extensions/amp-truncate-text/0.1/amp-truncate-text.js

```diff
@@ -0,0 +1,259 @@
+    if (sizer) {
+      sizer.setAttribute('slot', 'sizer');
+    }
+
+    // TODO(sparhami) Is there a shared place to add logic for creating
+    // shadow roots with styles? Might make sense to have it create the style
+    // as well as a slot for the sizer.
+    const sr = this.element.attachShadow({mode: 'open'});
+    const style = document.createElement('style');
+    style.textContent = ShadowCSS;
+    const html = htmlFor(this.element);
+    const content = html`
+      <div>
+        <div class="content">
+          <slot></slot>
+          <slot class="expand-slot" name="expand"></slot>
+          <slot class="collapse-slot" name="collapse"></slot>
+        </div>
+        <slot name="sizer"></slot>
+      </div>
+    `;
+
+    sr.appendChild(style);
+    sr.appendChild(content);
+
+    this.content_ = null;
+    this.expandSlot_ = content.querySelector('.expand-slot');
+    this.collapseSlot_ = content.querySelector('.collapse-slot');
+  }
+
+  /** @override */
+  layoutCallback() {
+    if (isExperimentOn(this.win, 'amp-truncate-text')) {
+      this.truncate_();
+    }
+
+    return Promise.resolve();
+  }
+
+  /** @override */
+  isRelayoutNeeded() {
+    return true;
+  }
+
+  /** @override */
+  isLayoutSupported() {
+    return true;
+  }
+
+  /**
+   * Stops listening for mutations, if we have a `MutationObserver`.
+   * @private
+   */
+  unlistenForMutations_() {
+    if (this.mutationObserver_) {
+      this.mutationObserver_.disconnect();
+    }
+  }
+
+  /**
+   * Starts listening for mutations, if we have a `MutationObserver`.
+   * @private
+   */
+  listenForMutations_() {
+    if (this.mutationObserver_) {
+      this.mutationObserver_.observe(this.element, {
+        attributes: true,
+        characterData: true,
+        childList: true,
+        subtree: true,
+      });
+    }
+  }
+
+  /**
+   * Truncates the content of the element. This is debounced as runtime will do
+   * a mutation (add a class) right after `layoutCallback`. We want to make
+   * sure we do not truncate twice as a result. TODO(sparhami) Try to find a
+   * better solution for this.
+   * @private
+   */
+  truncate_() {
+    if (this.truncateRequested_) {
+      return;
+    }
+
+    this.truncateRequested_ = true;
+
+    Promise.resolve()
+        .then(() => {
+        // Runtime will actually trigger another mutation one microtask after
+        // layoutCallback finishes, so we need to delay one additional
+        // microtask for that.
+        })
+        .then(() => {
+          this.truncateRequested_ = false;
+
+          const container = dev().assertElement(
+              this.useShadow_ ? this.element : this.content_);
+          const overflowElement = this.useShadow_ ?
+            this.element.querySelector('[slot="expand"]') :
+            this.element.querySelector('.i-amphtml-truncate-expand-slot');
+
+          // Make sure mutations from truncateing do not trigger truncateing.
+          this.unlistenForMutations_();
```

Comment:
How well do switching unlisten/listen perform?

Reply:
Another q: does `takeRecords` work for this to purge the queue?

Reply:
Yep, changed to use `takeRecords` instead, since that makes things more clean.

Reply:
Cool. Let's see how goes it. TBH, no idea what's better yet - would like to see what works.